staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -149,7 +163,20 @@ type SharedInformer interface {
 	// between any two resyncs may be longer than the nominal period
 	// because the implementation takes time to do work and there may
 	// be competing load and scheduling noise.
-	AddEventHandlerWithResyncPeriod(handler ResourceEventHandler, resyncPeriod time.Duration)
+	// It returns a handle for the handler that can be used to remove
+	// the handler again and an error if the handler cannot be added.
+	AddEventHandlerWithResyncPeriod(handler ResourceEventHandler, resyncPeriod time.Duration) (*ResourceEventHandlerHandle, error)
+	// RemoveEventHandlerByHandle removes a formerly added event handler given by
+	// its registration handle.
+	// If, for some reason, the same handler has been added multiple
+	// times, only the registration for the given registration handle
```

Comment:
shouldn't each time get a unique handle?

Reply:
The handle is unique for a registration, not for the handler, you can still add a handler multiple times, and remove the registrations one by one. It would be dangerous to enforce the removal of a handler registered multiple times, if only one registration is cancelled. 

The reason for introducing a handler was that handlers might not be comparable. Because of this it is also no possible to provide a handle unique for a handler.

Reply:
I'm having trouble parsing what you mean by "registration" and "handler".

Suppose I call this N times, passing ResourceEventHandler `h` each time. I would then expect that when a matching event arrives, `h` is notified N times. I would further expect that each call to AddEventHandlerWithResyncPeriod produced a different ResourceEventHandlerHandle and I would have to call remove for each handler handle to stop getting notifications.

I'd further expect that if I add `h` and someone else adds handler `g` which happens to match the same events, that when I remove my ResourceEventHandlerHandle, `g` would not be affected.

If it doesn't work like this, please make it work like this.

Please update the comment to make this clearer.

Reply:
Exactly this is how is works. With every add call a new new registration handle for the handler is created, which results in an additional listener, which uses the same handler. Therefore it is called n times, if it is added n times. The same is true for the deletion; the remove call for a handler just removes the listener associated with the handle wich refers to the originally given handler. Therefore there will be a decrementing amount of calls with every remove operation for an not already removed handle originally created for the same handler.